python/ray/tune/hpo_scheduler.py

```diff
@@ -0,0 +1,201 @@
+                self.rstate.randint(2 ** 31 - 1))
+            self._hpopt_trials.insert_trial_docs(new_trials)
+            self._hpopt_trials.refresh()
+            new_trial = new_trials[0]
+            new_trial_id = new_trial["tid"]
+            suggested_config = hpo.base.spec_from_misc(new_trial["misc"])
+            new_cfg.update(suggested_config)
+
+            kv_str = "_".join(["{}={}".format(k, str(v)[:5])
+                               for k, v in suggested_config.items()])
+            experiment_tag = "hyperopt_{}_{}".format(new_trial_id, kv_str)
+
+            # Keep this consistent with tune.variant_generator
+            trial = Trial(
+                trainable_name=self.args.run,
+                config=new_cfg,
+                local_dir=os.path.join(self.args.local_dir, self._output_path),
+                experiment_tag=experiment_tag,
+                resources=self.args.resources,
+                stopping_criterion=self.args.stop,
+                checkpoint_freq=self.args.checkpoint_freq,
+                restore_path=self.args.restore,
+                upload_dir=self.args.upload_dir,
+                max_failures=self.args.max_failures)
+
+            self._tune_to_hp[trial] = new_trial_id
+            self._num_trials_left -= 1
+            yield trial
+
+    def on_trial_result(self, trial_runner, trial, result):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        now = hpo.utils.coarse_utcnow()
+        ho_trial['book_time'] = now
+        ho_trial['refresh_time'] = now
+        return TrialScheduler.CONTINUE
+
+    def on_trial_error(self, trial_runner, trial):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_ERROR
+        ho_trial['misc']['error'] = (str(TuneError), "Tune Error")
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def on_trial_remove(self, trial_runner, trial):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_ERROR
+        ho_trial['misc']['error'] = (str(TuneError), "Tune Removed")
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def on_trial_complete(self, trial_runner, trial, result):
+        ho_trial = self._get_dynamic_trial(self._tune_to_hp[trial])
+        ho_trial['refresh_time'] = hpo.utils.coarse_utcnow()
+        ho_trial['state'] = hpo.base.JOB_STATE_DONE
+        hp_result = self._convert_result(result)
+        ho_trial['result'] = hp_result
+        self._hpopt_trials.refresh()
+        del self._tune_to_hp[trial]
+
+    def _convert_result(self, result):
+        return {"loss": getattr(result, self._loss_attr),
+                "status": "ok"}
+
+    def _get_dynamic_trial(self, tid):
```

Comment:
_get_hyperopt_trial

Also can't we do a dict lookup here?

Reply:
I would prefer not to since hyperopt is pretty fragile (this exact trial dictionary that we get from this method is not a copy but rather the actual object used in trial generation)

It wouldn't be horrible + we can write a test but one could imagine it as a source for error